executor/simple.go

```diff
@@ -566,13 +569,16 @@ func (e *SimpleExec) executeUse(s *ast.UseStmt) error {
 }
 
 func (e *SimpleExec) executeBegin(ctx context.Context, s *ast.BeginStmt) error {
-	// If `START TRANSACTION READ ONLY WITH TIMESTAMP BOUND` is the first statement in TxnCtx, we should
+	// If `START TRANSACTION READ ONLY` is the first statement in TxnCtx, we should
 	// always create a new Txn instead of reusing it.
 	if s.ReadOnly {
 		enableNoopFuncs := e.ctx.GetSessionVars().EnableNoopFuncs
-		if !enableNoopFuncs && s.Bound == nil {
+		if !enableNoopFuncs && s.AsOf == nil && s.Bound == nil {
 			return expression.ErrFunctionsNoopImpl.GenWithStackByArgs("READ ONLY")
 		}
+		if s.AsOf != nil {
+			return e.executeStartTransactionReadOnlyWithBoundedStaleness(ctx, s)
+		}
 		if s.Bound != nil {
 			return e.executeStartTransactionReadOnlyWithTimestampBound(ctx, s)
 		}
```

Comment:
I think this could be removed?

Reply:
Code cleaning work will be in an independent PR later.